Source/WebCore/platform/network/HTTPParsers.cpp

```diff
@@ -579,94 +579,259 @@ XFrameOptionsDisposition parseXFrameOptionsHeader(StringView header)
-                    if (header[index] == '\\') {
-                        ++index;
-                        if (index == header.length())
-                            return std::nullopt;
-                        if (header[index] != '\\' && header[index] != '"')
-                            return std::nullopt;
-                        valueBuilder.append(header[index]);
-                    } else if (header[index] == '\"') {
-                        value = valueBuilder.toString();
-                        break;
-                    } else if (header[index] <= 0x1F || (header[index] >= 0x7F && header[index] <= 0xFF)) // Not in VCHAR or SP.
-                        return std::nullopt;
-                    else
-                        valueBuilder.append(header[index]);
-                    ++index;
-                }
-                if (index == header.length())
-                    return std::nullopt;
-                ++index; // Skip DQUOTE.
-            } else
+            auto parsedMember = parseItemOrInnerList(string, index);
+            if (!parsedMember)
+                return std::nullopt;
+            member = WTFMove(*parsedMember);
+        } else {
+            BareItem value = true;
+            auto parameters = parseParameters(string, index);
+            if (!parameters)
                 return std::nullopt;
+            member = std::pair { WTFMove(value), WTFMove(*parameters) };
         }
-        parameters.set(key.toString(), WTFMove(value));
+        dictionary.set(key.toString(), WTFMove(member));
+        while (index < string.length() && RFC7230::isWhitespace(string[index]))
+            ++index;
+        if (index >= string.length())
+            return dictionary;
+        UChar c = string[index++];
+        if (c != ',')
+            return std::nullopt;
+        while (index < string.length() && RFC7230::isWhitespace(string[index]))
+            ++index;
+        if (index >= string.length())
+            return std::nullopt;
     }
-    if (index != header.length())
+    ASSERT(dictionary.isEmpty());
+    return dictionary;
+}
+
+// https://datatracker.ietf.org/doc/html/rfc8941#section-4.2 with type "item".
+std::optional<std::pair<BareItem, Parameters>> parseItemStructuredFieldValue(StringView header)
+{
+    if (header.isEmpty())
+        return std::nullopt;
+
+    size_t index = 0;
+    while (index < header.length() && header[index] == ' ')
+        ++index;
+
+    auto item = parseItem(header, index);
+    if (!item)
+        return std::nullopt;
+
+    while (index < header.length() && header[index] == ' ')
+        ++index;
+
+    if (index < header.length())
         return std::nullopt;
-    return std::make_pair(bareItem, parameters);
+    return WTFMove(*item);
 }
 
+// https://datatracker.ietf.org/doc/html/rfc8941#section-4.2 with type "dictionary".
+std::optional<HashMap<String, std::pair<ItemOrInnerList, Parameters>>> parseDictionaryStructuredFieldValue(StringView header)
```

Comment:
For new parsing code on top of our UTF16/Latin1 WTF::String classes, I have been trying to encourage use of WTF::StringParsingBuffer as reduces the overall branching of the parser by doing the UChar/LChar choice once up front.

Reply:
Paired with helpers like skipWhile<> and friends, it makes a pretty nice little parsing DSL.

Reply:
Ok, I am not familiar with StringParsingBuffer but I'll look into it.

Reply:
Do you mind if I do this in a follow-up? Seems like this change will be significant enough to require another review.